wfe2/verify.go

```diff
@@ -5,246 +5,540 @@ import (
+	// If the signature verification fails, then return an error immediately with
+	// a small bit of context from the JWS body
 	if err != nil {
-		wfe.stats.Inc("Errors.JWSVerificationFailed", 1)
 		n := len(body)
 		if n > 100 {
 			n = 100
 		}
+		wfe.stats.Inc("Errors.JWSVerificationFailed", 1)
 		logEvent.AddError("verification of JWS with the JWK failed: %v; body: %s", err, body[:n])
-		return nil, nil, reg, probs.Malformed("JWS verification error")
+		return nil, probs.Malformed("JWS verification error")
 	}
+	// Store the verified payload in the logEvent
 	logEvent.Payload = string(payload)
 
-	// Check that the request has a known anti-replay nonce
-	nonce := parsedJws.Signatures[0].Header.Nonce
-	logEvent.RequestNonce = nonce
-	if len(nonce) == 0 {
-		wfe.stats.Inc("Errors.JWSMissingNonce", 1)
-		logEvent.AddError("JWS is missing an anti-replay nonce")
-		return nil, nil, reg, probs.BadNonce("JWS has no anti-replay nonce")
-	} else if !wfe.nonceService.Valid(nonce) {
-		wfe.stats.Inc("Errors.JWSInvalidNonce", 1)
-		logEvent.AddError("JWS has an invalid anti-replay nonce: %s", nonce)
-		return nil, nil, reg, probs.BadNonce(fmt.Sprintf("JWS has invalid anti-replay nonce %v", nonce))
+	// Check that the JWS contains a correct Nonce header
+	if prob := wfe.validNonce(jws, logEvent); prob != nil {
+		return nil, prob
 	}
 
-	// Check that the "resource" field is present and has the correct value
-	var parsedRequest struct {
-		Resource string `json:"resource"`
+	// Check that the HTTP request URL matches the URL in the signed JWS
+	if prob := wfe.validPOSTURL(request, jws, logEvent); prob != nil {
+		return nil, prob
 	}
-	err = json.Unmarshal([]byte(payload), &parsedRequest)
-	if err != nil {
+
+	// Previously the check for the request URL required unmarshalling the payload JSON
+	// to check the "resource" field of the protected JWS body. This caught
+	// invalid JSON early and so we preserve this check by explicitly trying to
+	// unmarshal the payload as part of the verification and failing early if it
+	// isn't JSON.
+	var parsedBody struct{}
+	if err := json.Unmarshal(payload, &parsedBody); err != nil {
 		wfe.stats.Inc("Errors.UnparseableJWSPayload", 1)
-		logEvent.AddError("unable to JSON parse resource from JWS payload: %s", err)
-		return nil, nil, reg, probs.Malformed("Request payload did not parse as JSON")
+		logEvent.AddError("POST JWS Body is invalid JSON: %q", err.Error())
+		return nil, probs.Malformed("Request payload did not parse as JSON")
+	}
+
+	return payload, nil
+}
+
+// validPOSTForAccount checks that a given POST request has a valid JWS
+// verified with the public key associated to a known account,
+// specified by the JWS key ID. If the request is valid (e.g. the JWS is well
+// formed, verifies with the JWK stored for the specified key ID, specifies the
+// correct URL, and has a valid nonce) then `validPOSTForAccount` returns the
+// validated JWS body, the JWK used to validate the JWS, the JWS that was
```

Comment:
I think I'm in favour of taking this review feedback and removing the return values not currently being used. When we implement keyrollover & revocation we can adjust the return signatures if it is truly required. I needed the return value based on my sketchy WIP implementation but there's been pretty significant refactoring since then.